File: src/test/java/com/zepben/blobstore/itemwrappers/ByDateItemReaderTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.blobstore.itemwrappers;

import com.zepben.blobstore.BlobStoreException;
import com.zepben.blobstore.WhereBlob;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatcher;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.Spy;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.*;

import static com.zepben.blobstore.CustomMatchers.eqArrayValueMap;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

public class ByDateItemReaderTest {

    private Set<String> ids = new HashSet<>(Arrays.asList("id1", "id2"));
    private static Set<String> tags = new HashSet<>(Arrays.asList("tag1", "tag2"));

    @Spy private MockBlobReader blobReader = new MockBlobReader(ids, tags);
    @Mock private ByDateBlobReaderProvider blobReaderProvider;
    @Mock private ByDateTagDeserialiser tag1Deserialiser;
    @Mock private ByDateTagDeserialiser tag2Deserialiser;
    @Mock private ByDateItemDeserialiser<String> itemDeserialiser;
    @Mock private ByDateItemError itemError;
    @Mock private ByDateItemHandler<String> itemHandler;

    private ZoneId timeZone = ZoneId.systemDefault();
    private ByDateItemReader<String> itemReader;

    @BeforeEach
    public void before() throws Exception {
        MockitoAnnotations.initMocks(this);

        doReturn(blobReader).when(blobReaderProvider).getReader(any(), any());

        Map<String, ByDateTagDeserialiser<?>> tagDeserialisers = new HashMap<>();
        tagDeserialisers.put("tag1", tag1Deserialiser);
        tagDeserialisers.put("tag2", tag2Deserialiser);
        itemReader = new ByDateItemReader<>(timeZone, blobReaderProvider);
        itemReader.setDeserialisers(itemDeserialiser, tagDeserialisers);
    }

    @Test
    public void get() throws Exception {
        String id = "id1";
        LocalDate date = LocalDate.now();

        itemReader = spy(itemReader);
        doReturn("theItem").when(itemDeserialiser).deserialise(any(), any(), any());

        String result = itemReader.get(id, date, itemError);
        verify(itemReader).forEach(eq(Collections.singleton(id)), eq(date), any(), eq(itemError));
        assertThat(result, equalTo("theItem"));
    }

    @Test
    public void getNullBlobReaderProvider() throws Exception {
        String id = "id1";
        LocalDate date = LocalDate.now();

        doReturn(null).when(blobReaderProvider).getReader(any(), any());
        itemReader = spy(itemReader);
        String result = itemReader.get(id, date, itemError);
        assertThat(result, is(nullValue()));
    }

    @Test
    public void forEach() throws Exception {
        LocalDate date = LocalDate.now();
        List<String> ids = Collections.singletonList("id1");

        doReturn("theItem").when(itemDeserialiser).deserialise(any(), any(), any());
        itemReader.forEach(ids, date, itemHandler, itemError);
        verify(blobReader).forEach(eq(ids), eq(tags), any());

        verify(blobReaderProvider).getReader(date, timeZone);
        ArgumentMatcher<Map<String, byte[]>> matchesMap = eqArrayValueMap(blobReader.getBlobs(tags));
        verify(itemDeserialiser).deserialise(eq("id1"), eq(date), argThat(matchesMap));
        verify(itemHandler).handle("id1", date, "theItem");
    }

    @Test
    public void forEachNullBlobReaderProvider() throws Exception {
        LocalDate date = LocalDate.now();
        List<String> ids = Collections.singletonList("id1");

        doReturn(null).when(blobReaderProvider).getReader(any(), any());
        itemReader.forEach(ids, date, itemHandler, itemError);
        verify(blobReaderProvider).getReader(date, timeZone);
    }

    @Test
    public void forAll() throws Exception {
        LocalDate date = LocalDate.now();

        doReturn("theItem1").when(itemDeserialiser).deserialise(eq("id1"), any(), any());
        doReturn("theItem2").when(itemDeserialiser).deserialise(eq("id2"), any(), any());
        itemReader.forAll(date, itemHandler, itemError);
        verify(blobReader).forAll(eq(tags), any());

        verify(blobReaderProvider).getReader(date, timeZone);
        ArgumentMatcher<Map<String, byte[]>> matchesMap = eqArrayValueMap(blobReader.getBlobs(tags));
        verify(itemDeserialiser).deserialise(eq("id1"), eq(date), argThat(matchesMap));
        verify(itemDeserialiser).deserialise(eq("id2"), eq(date), argThat(matchesMap));
        verify(itemHandler).handle("id1", date, "theItem1");
        verify(itemHandler).handle("id2", date, "theItem2");
    }

    @Test
    public void forAllWhere() throws Exception {
        LocalDate date = LocalDate.now();
        List<WhereBlob> whereBlobs = Collections.singletonList(WhereBlob.Companion.equals("tag1", new byte[]{}));

        doReturn("theItem").when(itemDeserialiser).deserialise(any(), any(), any());
        itemReader.forAll(date, whereBlobs, itemHandler, itemError);
        verify(blobReader).forAll(eq(tags), eq(whereBlobs), any());

        verify(blobReaderProvider).getReader(date, timeZone);
        ArgumentMatcher<Map<String, byte[]>> matchesMap = eqArrayValueMap(blobReader.getBlobs(tags));
        verify(itemDeserialiser).deserialise(eq("id1"), eq(date), argThat(matchesMap));
        verify(itemHandler).handle("id1", date, "theItem");
    }

    @Test
    public void forAllNullBlobReaderProvider() throws Exception {
        LocalDate date = LocalDate.now();

        doReturn(null).when(blobReaderProvider).getReader(any(), any());
        itemReader.forAll(date, itemHandler, itemError);
        verify(blobReaderProvider).getReader(date, timeZone);
    }

    @Test
    public void handlesBlobReaderException() throws BlobStoreException {
        Set<String> ids = Collections.singleton("id");
        LocalDate date = LocalDate.now();

        BlobStoreException ex = new BlobStoreException("test", null);
        doAnswer(inv -> {throw ex;}).when(blobReader).forEach(anyCollection(), anySet(), any());
        doAnswer(inv -> {throw ex;}).when(blobReader).forAll(anySet(), any());

        itemReader.forEach(ids, date, itemHandler, itemError);
        itemReader.forAll(date, itemHandler, itemError);
        verify(itemError, times(2)).handle(eq(""), eq(date), anyString(), eq(ex));
    }

    @Test
    public void getTag() throws Exception {
        String id = "id1";
        String tag = "tag1";
        LocalDate date = LocalDate.now();

        itemReader = spy(itemReader);
        doReturn("theTag").when(tag1Deserialiser).deserialise(any(), any(), any(), any());
        String result = itemReader.get(id, date, tag, itemError);

        verify(blobReader).get(id, tag);
        byte[] expectedBytes = blobReader.getBlob(tag);
        verify(tag1Deserialiser).deserialise(eq(id), eq(date), eq(tag), eq(expectedBytes));
        assertThat(result, equalTo("theTag"));
    }

    @Test
    public void getTagNullBlobReaderProvider() throws Exception {
        String id = "id1";
        String tag = "tag1";
        LocalDate date = LocalDate.now();

        doReturn(null).when(blobReaderProvider).getReader(any(), any());
        String result = itemReader.get(id, date, tag, itemError);
        assertThat(result, is(nullValue()));
    }

    @Test
    public void forEachTag() throws Exception {
        LocalDate date = LocalDate.now();
        List<String> ids = Collections.singletonList("id1");
        String tag = "tag1";

        doReturn("theTag").when(tag1Deserialiser).deserialise(any(), any(), any(), any());
        itemReader.forEach(ids, date, tag, itemHandler, itemError);
        verify(blobReader).forEach(eq(ids), eq(tag), any());

        verify(blobReaderProvider).getReader(date, timeZone);
        byte[] expectedBytes = blobReader.getBlob(tag);
        verify(tag1Deserialiser).deserialise(eq("id1"), eq(date), eq(tag), eq(expectedBytes));
        verify(itemHandler).handle("id1", date, "theTag");
    }

    @Test
    public void forEachTagNullBlobReaderProvider() throws Exception {
        LocalDate date = LocalDate.now();
        List<String> ids = Collections.singletonList("id1");
        String tag = "tag1";

        doReturn(null).when(blobReaderProvider).getReader(any(), any());
        itemReader.forEach(ids, date, tag, itemHandler, itemError);
        verify(blobReaderProvider).getReader(date, timeZone);
    }

    @Test
    public void forAllTag() throws Exception {
        LocalDate date = LocalDate.now();
        String tag = "tag1";

        doReturn("theTag").when(tag1Deserialiser).deserialise(any(), any(), any(), any());
        itemReader.forAll(date, tag, itemHandler, itemError);
        verify(blobReader).forAll(eq(tag), any());

        verify(blobReaderProvider).getReader(date, timeZone);
        byte[] expectedBytes = blobReader.getBlob(tag);
        verify(tag1Deserialiser).deserialise(eq("id1"), eq(date), eq(tag), eq(expectedBytes));
        verify(tag1Deserialiser).deserialise(eq("id2"), eq(date), eq(tag), eq(expectedBytes));
        verify(itemHandler).handle("id1", date, "theTag");
        verify(itemHandler).handle("id2", date, "theTag");
    }

    @Test
    public void forAllTagNullBlobReaderProvider() throws Exception {
        LocalDate date = LocalDate.now();
        String tag = "tag1";

        doReturn(null).when(blobReaderProvider).getReader(any(), any());
        itemReader.forAll(date, tag, itemHandler, itemError);
        verify(blobReaderProvider).getReader(date, timeZone);
    }

    @Test
    public void getBlobReaderHandlesException() throws BlobStoreException {
        BlobStoreException ex = new BlobStoreException("test", null);
        when(blobReaderProvider.getReader(any(), any())).thenThrow(ex);

        LocalDate date = LocalDate.now();
        itemReader.forEach(Collections.singleton("id"), date, itemHandler, itemError);
        itemReader.forAll(date, itemHandler, itemError);

        itemReader.forEach(Collections.singleton("id"), date, "tag1", itemHandler, itemError);
        itemReader.forAll(date, "tag1", itemHandler, itemError);

        verify(itemError, times(4)).handle(eq(""), eq(date), anyString(), eq(ex));
    }

    @Test
    public void handlesItemDeserialiseException() throws Exception {
        String id = "id1";
        LocalDate date = LocalDate.now();

        DeserialiseException ex = new DeserialiseException("test", null);
        doThrow(ex).when(itemDeserialiser).deserialise(any(), any(), any());
        String str = itemReader.get(id, date, itemError);
        assertNull(str);
        verify(itemError).handle(id, date, "test", ex);
    }

    @Test
    public void handlesTagDeserialiseException() throws Exception {
        String id = "id1";
        LocalDate date = LocalDate.now();

        DeserialiseException ex = new DeserialiseException("test", null);
        doThrow(ex).when(tag1Deserialiser).deserialise(any(), any(), any(), any());
        String str = itemReader.get(id, date, "tag1", itemError);
        assertNull(str);
        verify(itemError).handle(id, date, "test", ex);
    }

}

File: src/test/kotlin/com/zepben/blobstore/itemwrappers/ByDateItemReaderTest.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.blobstore.itemwrappers

import com.zepben.blobstore.BlobStoreException
import com.zepben.blobstore.CustomMatchers.eqArrayValueMap
import com.zepben.blobstore.WhereBlob
import io.mockk.*
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.nullValue
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.ZoneId

class ByDateItemReaderTest {

    private val ids: Set<String> = HashSet(mutableListOf("id1", "id2"))
    private val tags: Set<String> = HashSet(mutableListOf("tag1", "tag2"))

    private val blobReader = spyk(MockBlobReader(ids, tags))

    private val blobReaderProvider = mockk<ByDateBlobReaderProvider>() { every { getReader(any(), any()) } returns blobReader }

    private val tag1Deserialiser = mockk<ByDateTagDeserialiser<*>>()
    private val tag2Deserialiser = mockk<ByDateTagDeserialiser<*>>()
    private val tagDeserialisers = mapOf("tag1" to tag1Deserialiser, "tag2" to tag2Deserialiser)

    private val itemDeserialiser = mockk<ByDateItemDeserialiser<String>>()
    private val itemError = mockk<ByDateItemError>()
    private val itemHandler = mockk<ByDateItemHandler<String>> { justRun { handle(any(), any(), any()) } }

    private val timeZone = ZoneId.systemDefault()
    private val date = LocalDate.now()

    private val itemReader = spyk(ByDateItemReader<String>(timeZone, blobReaderProvider).apply {
        setDeserialisers(itemDeserialiser, tagDeserialisers)
    })

    @Test
    fun get() {
        val id = "id1"

        every { itemDeserialiser.deserialise(any(), any(), any()) } returns "theItem"

        val result = itemReader[id, date, itemError]

        assertThat(result, equalTo("theItem"))
        verify { itemReader.forEach(setOf(id), date, any(), itemError) }
    }

    @Test
    fun getNullBlobReaderProvider() {
        val id = "id1"

        every { blobReaderProvider.getReader(any(), any()) } returns null

        val result = itemReader[id, date, itemError]

        assertThat(result, nullValue())
    }

    @Test
    fun forEach() {
        val ids = listOf("id1")
        val expectedBytes = blobReader.getBlobs(tags).also { clearMocks(blobReader, answers = false) }

        every { itemDeserialiser.deserialise(any(), any(), any()) } returns "theItem"

        itemReader.forEach(ids, date, itemHandler, itemError)

        verify { blobReader.forEach(ids, tags, any()) }
        verify { blobReaderProvider.getReader(date, timeZone) }
        verify { itemDeserialiser.deserialise("id1", date, eqArrayValueMap(expectedBytes)) }
        verify { itemHandler.handle("id1", date, "theItem") }
    }

    @Test
    fun forEachNullBlobReaderProvider() {
        val ids = listOf("id1")

        every { blobReaderProvider.getReader(any(), any()) } returns null

        itemReader.forEach(ids, date, itemHandler, itemError)

        verify { blobReaderProvider.getReader(date, timeZone) }
    }

    @Test
    fun forAll() {
        val expectedBytes = blobReader.getBlobs(tags).also { clearMocks(blobReader, answers = false) }

        every { itemDeserialiser.deserialise("id1", any(), any()) } returns "theItem1"
        every { itemDeserialiser.deserialise("id2", any(), any()) } returns "theItem2"

        itemReader.forAll(date, itemHandler, itemError)

        verify { blobReader.forAll(tags, any()) }
        verify { blobReaderProvider.getReader(date, timeZone) }
        verify { itemDeserialiser.deserialise("id1", date, eqArrayValueMap(expectedBytes)) }
        verify { itemDeserialiser.deserialise("id2", date, eqArrayValueMap(expectedBytes)) }
        verify { itemHandler.handle("id1", date, "theItem1") }
        verify { itemHandler.handle("id2", date, "theItem2") }
    }

    @Test
    fun forAllWhere() {
        val whereBlobs = listOf(WhereBlob.equals("tag1", byteArrayOf()))
        val expectedBytes = blobReader.getBlobs(tags).also { clearMocks(blobReader, answers = false) }

        every { itemDeserialiser.deserialise(any(), any(), any()) } returns "theItem"

        itemReader.forAll(date, whereBlobs, itemHandler, itemError)

        verify { blobReader.forAll(tags, whereBlobs, any()) }
        verify { blobReaderProvider.getReader(date, timeZone) }
        verify { itemDeserialiser.deserialise("id1", date, eqArrayValueMap(expectedBytes)) }
        verify { itemHandler.handle("id1", date, "theItem") }
    }

    @Test
    fun forAllNullBlobReaderProvider() {

        every { blobReaderProvider.getReader(any(), any()) } returns null

        itemReader.forAll(date, itemHandler, itemError)

        verify { blobReaderProvider.getReader(date, timeZone) }
    }

    @Test
    fun handlesBlobReaderException() {
        val ids = setOf("id")
        val ex = BlobStoreException("test", null)

        every { blobReader.forEach(any(), any<Set<String>>(), any()) } throws ex
        every { blobReader.forAll(any<Set<String>>(), any()) } throws ex
        justRun { itemError.handle(any(), any(), any(), any()) }

        itemReader.forEach(ids, date, itemHandler, itemError)
        itemReader.forAll(date, itemHandler, itemError)

        verify(exactly = 2) { itemError.handle("", date, any(), ex) }
    }

    @Test
    fun getTag() {
        val id = "id1"
        val tag = "tag1"
        val expectedBytes = blobReader.getBlob(tag).also { clearMocks(blobReader, answers = false) }

        every { tag1Deserialiser.deserialise(any(), any(), any(), any()) } returns "theTag"

        val result = itemReader.get<String>(id, date, tag, itemError)

        assertThat(result, equalTo("theTag"))

        verify { blobReader[id, tag] }
        verify { tag1Deserialiser.deserialise(id, date, tag, expectedBytes) }
    }

    @Test
    fun getTagNullBlobReaderProvider() {
        val id = "id1"
        val tag = "tag1"

        every { blobReaderProvider.getReader(any(), any()) } returns null

        val result = itemReader.get<String>(id, date, tag, itemError)

        assertThat(result, nullValue())
    }

    @Test
    fun forEachTag() {
        val ids = listOf("id1")
        val tag = "tag1"
        val expectedBytes = blobReader.getBlob(tag).also { clearMocks(blobReader, answers = false) }

        every { tag1Deserialiser.deserialise(any(), any(), any(), any()) } returns "theTag"

        itemReader.forEach(ids, date, tag, itemHandler, itemError)

        verify { blobReader.forEach(ids, tag, any()) }
        verify { blobReaderProvider.getReader(date, timeZone) }
        verify { tag1Deserialiser.deserialise("id1", date, tag, expectedBytes) }
        verify { itemHandler.handle("id1", date, "theTag") }
    }

    @Test
    fun forEachTagNullBlobReaderProvider() {
        val ids = listOf("id1")
        val tag = "tag1"

        every { blobReaderProvider.getReader(any(), any()) } returns null

        itemReader.forEach(ids, date, tag, itemHandler, itemError)

        verify { blobReaderProvider.getReader(date, timeZone) }
    }

    @Test
    fun forAllTag() {
        val tag = "tag1"
        val expectedBytes = blobReader.getBlob(tag).also { clearMocks(blobReader, answers = false) }

        every { tag1Deserialiser.deserialise(any(), any(), any(), any()) } returns "theTag"

        itemReader.forAll(date, tag, itemHandler, itemError)

        verify { blobReader.forAll(tag, any()) }
        verify { blobReaderProvider.getReader(date, timeZone) }
        verify { tag1Deserialiser.deserialise("id1", date, tag, expectedBytes) }
        verify { tag1Deserialiser.deserialise("id2", date, tag, expectedBytes) }
        verify { itemHandler.handle("id1", date, "theTag") }
        verify { itemHandler.handle("id2", date, "theTag") }
    }

    @Test
    fun forAllTagNullBlobReaderProvider() {
        val tag = "tag1"

        every { blobReaderProvider.getReader(any(), any()) } returns null

        itemReader.forAll(date, tag, itemHandler, itemError)

        verify { blobReaderProvider.getReader(date, timeZone) }
    }

    @Test
    fun getBlobReaderHandlesException() {
        val ex = BlobStoreException("test", null)

        every { blobReaderProvider.getReader(any(), any()) } throws ex
        justRun { itemError.handle(any(), any(), any(), any()) }

        itemReader.forEach(setOf("id"), date, itemHandler, itemError)
        itemReader.forAll(date, itemHandler, itemError)
        itemReader.forEach(setOf("id"), date, "tag1", itemHandler, itemError)
        itemReader.forAll(date, "tag1", itemHandler, itemError)

        verify(exactly = 4) { itemError.handle("", date, any(), ex) }
    }

    @Test
    fun handlesItemDeserialiseException() {
        val id = "id1"
        val ex = DeserialiseException("test", null)

        every { itemDeserialiser.deserialise(any(), any(), any()) } throws ex
        justRun { itemError.handle(any(), any(), any(), any()) }

        val str = itemReader[id, date, itemError]

        assertThat(str, nullValue())
        verify { itemError.handle(id, date, "test", ex) }
    }

    @Test
    fun handlesTagDeserialiseException() {
        val id = "id1"
        val ex = DeserialiseException("test", null)

        every { tag1Deserialiser.deserialise(any(), any(), any(), any()) } throws ex
        justRun { itemError.handle(any(), any(), any(), any()) }

        val str = itemReader.get<String>(id, date, "tag1", itemError)

        assertThat(str, nullValue())
        verify { itemError.handle(id, date, "test", ex) }
    }

}
